source_gen/lib/src/output_helpers.dart

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:async/async.dart';
+
+/// Runs [run] and converts [Future], [Iterable], and [Stream] implementations
+/// containing [String] to a single [Stream] while ensuring all thrown
+/// exceptions are forwarded through the return value.
+Stream<String> normalizeGeneratorOutput(Object run()) => StreamCompleter
+    .fromFuture(new Future.sync(run).then(_normalizeGeneratorOutput));
+
+Stream<String> _normalizeGeneratorOutput(Object value) {
+  if (value is String) {
+    value = [value];
+  }
+
+  value ??= [];
+
+  if (value is Iterable) {
+    value = safeIteratorToStream(value as Iterable);
+  }
+
+  if (value is Stream) {
```

Comment:
nit: check for is! stream so the error is closer to the code that is really checking for it?

Reply:
But then I have to cast the value I just checked – because we don't do inference for values after `is! Stream`.

Okay if I leave it?

Reply:
I mean, I get where you are going with that but it still makes the code harder to read imo. I don't care _that_ much though.

If only we had:

```
if (value is! Stream) {
  ...
} else {
  value is now a stream here...
} 
``

Reply:
agreed!